### PR TITLE
Add beer-song exercise

### LINKED ISSUE #49: Add "Beer Song" exercise

### PR DESCRIPTION
**Pull request by palisaide**
Closes #49

### CODE REVIEW

**Comment by patricksjackson** [comment]
I tested this and it works fine, but it needs at least one small tweak:

The example solution needs to be written as if the filenames are actually beer_song.c/h. So the include guards and the `#include` need to reference `beer_song.c/h`. This makes it easy to just rename the files for testing.

The function `resetTest()` in the testing file doesn't appear to be necessary, although I don't know all that much about Unity. At least some of the other problems have the same stub though...

(Formatting will need to change after we close #40, but we're going to need to make a pass through all the problems anyway, so I think this is fine for now)

**Comment by palisaide** [comment]
@patricksjackson, thanks for clarifying the example convention.  I saw some different approaches in the other exercises and I wasn't sure which was correct.  However, now I'm wondering if there's any reason to call the example `example.c` and not just `beer_song.c`.  I think it would be obvious that the files in `src` are just an example.  It would also make it easier to get the CI build to actually run the tests on the examples.

I just tested without the empty `resetTest()` and everything appears to be fine, so I guess Unity doesn't need it.

**Comment by patricksjackson** [comment]
They need to be called `example.*` so that the exercism client doesn't download them when you fetch exercises. If they were named `beer_song.*` they'd be downloaded along with the rest of the exercise. Doing it that way lets the tracks provide skeleton files if they want to.

Everything looks good to me :) :+1:

File: config.json [hunk 1]
```diff
@@ -74,6 +75,13 @@
       "bitwise operations",
       "performance optimizations"
     ]
+  }, {
+    "difficulty": 2,
+    "slug": "beer-song",
+    "topics": [
+      "strings",
+      "control-flow (if-else statements)"
```
Comment (ryanplusplus) [inline config.json]: How about `control-flow (loops)`?

File: exercises/beer-song/src/example.c [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+      "Take it down and pass it around, no more bottles of beer on the wall.\n");
+  } else if(verseNumber <= 99) {
+    bytesWritten = sprintf(buffer, "%d bottles of beer on the wall, %d bottles of beer.\n"
+      "Take one down and pass it around, %d bottles of beer on the wall.\n",
+      verseNumber, verseNumber, verseNumber - 1);
+  }
+
+  return bytesWritten;
+}
+void verse(char *buffer, int verseNumber) {
```
Comment (ryanplusplus) [inline exercises/beer-song/src/example.c]: Can you add a newline above this?

File: exercises/beer-song/test/test_beer_song.c [hunk 3]
```diff
@@ -0,0 +1,93 @@
...
+    "Go to the store and buy some more, 99 bottles of beer on the wall.\n";
+
+  sing(response, 3, 0);
+  TEST_ASSERT_EQUAL_STRING(expected, response);
+}
+
+int main(void) {
+  UnityBegin("test/test_beer_song.c");
+
+  RUN_TEST(test_handles_arbirtary_verse);
```
Comment (ryanplusplus) [inline exercises/beer-song/test/test_beer_song.c]: `arbirtrary` => `arbitrary`

File: exercises/beer-song/test/test_beer_song.c [hunk 4]
```diff
@@ -0,0 +1,93 @@
...
+    "6 bottles of beer on the wall, 6 bottles of beer.\n"
+    "Take one down and pass it around, 5 bottles of beer on the wall.\n";
+
+  sing(response, 8, 6);
+  TEST_ASSERT_EQUAL_STRING(expected, response);
+}
+
+void test_sings_the_last_4_verses(void) {
+  char response[BUFFER_SIZE * 4];
+  const char expected[BUFFER_SIZE * 4] =
```
Comment (ryanplusplus) [inline exercises/beer-song/test/test_beer_song.c]: Any reason to not just use the larger buffer always?

File: exercises/beer-song/src/example.c [hunk 5]
```diff
@@ -0,0 +1,40 @@
...
+
+  if(verseNumber == 0) {
+    bytesWritten = sprintf(buffer, "No more bottles of beer on the wall, no more bottles of beer.\n"
+      "Go to the store and buy some more, 99 bottles of beer on the wall.\n");
+  } else if(verseNumber == 2) {
+    bytesWritten = sprintf(buffer, "2 bottles of beer on the wall, 2 bottles of beer.\n"
+        "Take one down and pass it around, 1 bottle of beer on the wall.\n");
+  } else if(verseNumber == 1) {
+    bytesWritten = sprintf(buffer, "1 bottle of beer on the wall, 1 bottle of beer.\n"
+      "Take it down and pass it around, no more bottles of beer on the wall.\n");
```
Comment (ryanplusplus) [inline exercises/beer-song/src/example.c]: Pretty serious nitpick here: your cases go 0, 2, then 1. How about 0, 1, 2?

File: exercises/beer-song/src/example.c [hunk 6]
```diff
@@ -0,0 +1,40 @@
...
+
+static int getVerse(char *buffer, int verseNumber) {
+  int bytesWritten = 0;
+
+  if(verseNumber == 0) {
+    bytesWritten = sprintf(buffer, "No more bottles of beer on the wall, no more bottles of beer.\n"
+      "Go to the store and buy some more, 99 bottles of beer on the wall.\n");
+  } else if(verseNumber == 2) {
+    bytesWritten = sprintf(buffer, "2 bottles of beer on the wall, 2 bottles of beer.\n"
+        "Take one down and pass it around, 1 bottle of beer on the wall.\n");
```
Comment (ryanplusplus) [inline exercises/beer-song/src/example.c]: Indentation is a bit off here

File: exercises/beer-song/src/example.c [hunk 7]
```diff
@@ -0,0 +1,40 @@
...
+      verseNumber, verseNumber, verseNumber - 1);
+  }
+
+  return bytesWritten;
+}
+void verse(char *buffer, int verseNumber) {
+  (void)getVerse(buffer, verseNumber);
+}
+
+void sing(char *buffer, int startingVerseNumber, int endingVerseNumber) {
```
Comment (ryanplusplus) [inline exercises/beer-song/src/example.c]: Should we use an unsigned value for the verse numbers?

**Comment by ryanplusplus** [comment]
Looks good to me. Almost ready.

File: exercises/beer-song/src/example.c (hunk 7 above)
Comment (palisaide) [inline exercises/beer-song/src/example.c]: Yes.

File: exercises/beer-song/src/example.c (hunk 5 above)
Comment (palisaide) [inline exercises/beer-song/src/example.c]: Yep, this bothers me as well.

File: exercises/beer-song/src/example.h [hunk 8]
```diff
@@ -0,0 +1,7 @@
+#ifndef _BEER_SONG_H
+#define _BEER_SONG_H
+
+void verse(char *buffer, int verseNumber);
+void sing(char *buffer, int startingVerseNumber, int endingVerseNumber);
```
Comment (ryanplusplus) [inline exercises/beer-song/src/example.h]: Looks like the prototypes didn't get the `unsigned` treatment

File: exercises/beer-song/src/example.c [hunk 9]
```diff
@@ -0,0 +1,40 @@
...
+  }
+
+  return bytesWritten;
+}
+
+void verse(char *buffer, unsigned int verseNumber) {
+  (void)getVerse(buffer, verseNumber);
+}
+
+void sing(char *buffer, int startingVerseNumber, int endingVerseNumber) {
```
Comment (ryanplusplus) [inline exercises/beer-song/src/example.c]: These guys are still signed

**Comment by kytrinyx** [comment]
> The example solution needs to be written as if the filenames are actually 

@patricksjackson Would this be worth articulating in the README for the track?

**Comment by patricksjackson** [comment]
@kytrinyx Yep, it's something to call out. The filenaming scheme is common for all the tracks (so it's in the x-common contributing guide) but C and Cpp have to actually have this conflict in submitted code.

I'll make sure it gets in.

**Comment by patricksjackson** [comment]
LGTM

File: exercises/beer-song/makefile [hunk 10]
```diff
@@ -0,0 +1,26 @@
...
+CFLAGS += -Wold-style-definition
+
+
+test: tests.out
+	@./tests.out
+
+clean:
+	rm -f *.o *.out
+
+tests.out: test/test_beer_song.c src/beer_song.c
```
Comment (ryanplusplus) [inline exercises/beer-song/makefile]: Not to prolong this forever, but can you add `src/beer_song.h` to make sure this rebuilds if the header is changed?

**Comment by ryanplusplus** [comment]
Can you re-run the indent script as well? I'm fairly certain this passes Travis only because those checks were added later.

**Pull request merged**